### PR TITLE
dbAppendTable: do not check for existence twice

### PR DESCRIPTION
**Pull request by detule**
Hi - 

This is a performance enhancement when using `dbAppendTable`.

We have seen recently that calls to `dbExistsTable` can be costly.  Avoid calling it twice when using the `dbAppendTable` API.

### CODE REVIEW

**Comment by detule** [comment]
Not in love with the argument name or style.

Feedback welcome!

**Comment by hadley** [comment]
What if we just remove `stopifnot(dbExistsTable(conn, name))` from `dbAppendTable()`? If the user calls it incorrectly on a table that doesn't exist, they'll now get an error from the database rather than R, but I think that's fine. We might also be able to remove the explicit `dbExistsTable()` check in `odbc_write_table()` in a similar way, but I think that would require a bit more analysis. (And maybe a little wrapping with `withCallingHandlers()` to ensure the user is adequately informed about what's happening).

We might instead (or also?) consider moving the append part of `odbc_write_table()` into `dbAppendTable()` so that `dbWriteTable()` becomes mostly a wrapper around `dbExists()`/`dbRemoveTable()` + `dbCreateTable()` + `dbAppendTable()`.

**Comment by detule** [comment]
Hey Hadley, Simon:

I like the idea of letting the DB report the error - made the change, thanks.
I vote to tackle the dbWrite->dbAppend, rather than dbAppend->dbWrite task / refactor in a separate ticket - though happy to also tackle here.

**Review by hadley — APPROVED** [review]
I think it's fine to do the bigger change in separate PR, but you might want to note that this one fixes #480, and if you don't start work on the next change right away, make an issue so we don't forget about it.

**Review by simonpcouch — APPROVED** [review]
I'm on board!🙂

**Comment by hadley** [comment]
I'll merge this just so it's in the next release.

**Comment by detule** [comment]
Sorry about that - figured out why I went with the new argument to `odbc_write_table` to begin with - it's because `dbWriteTable`, per the DBI spec, should create the table on-the-fly if it does not exist ( unlike `dbAppendTable` where the expectation is that it fails ).

Given that the release is out the door, i might spend a bit of time to try and fit this within the `dbAppendTable`/`dbWriteTable` refactor we discussed.

**Comment by hadley** [comment]
No problems; hopefully the fix will fall in a fairly straightforward way once `dbWriteTable()` is implemented in terms of `dbCreateTable()` and `dbAppendTable()`.

**Comment by detule** [comment]
Thanks for the patience / gave it a shot.  Let me know if anything looks suspect.
Will re-run pipeline once upstream issues have been resolved.

File: R/dbi-table.R [hunk 1]
```diff
@@ -152,21 +117,71 @@ setMethod("dbWriteTable", c("OdbcConnection", "SQL", "data.frame"),
...
+#' chained together from needlessly iterating across the columns of
+#' the data.frame.  It is unlikely that users would ever need to set ( or
+#' interogate ) this attribute directly.
 #' @inheritParams DBI::dbReadTable
 #' @export
 setMethod("sqlData", "OdbcConnection",
   function(con, value, row.names = NA, ...) {
+
+    chk <- attr(value, ".odbc.transformed", exact = TRUE)
+    if ( !is.null(chk) && chk == TRUE ) return(value)
```
Comment (hadley) [inline R/dbi-table.R:183]:
```suggestion
    if (isTRUE(chk)) return(value)
```

File: R/dbi-table.R [hunk 2]
```diff
@@ -152,21 +117,71 @@ setMethod("dbWriteTable", c("OdbcConnection", "SQL", "data.frame"),
...
+      )
+    }
+
     invisible(NA_real_)
-  }
-)
+  })
 
 #' @rdname DBI-methods
+#' @details If the `value` argument has an attribute `.odbc.transformed` that
```
Comment (hadley) [inline R/dbi-table.R:171]: IMO since this is purely for internal use, you could omit from the docs.

File: R/dbi-table.R [hunk 3]
```diff
@@ -36,10 +36,11 @@ odbc_write_table <- function(conn,
                              overwrite = FALSE,
                              append = FALSE,
                              temporary = FALSE,
-                             row.names = NA,
+                             row.names = NULL,
```
Comment (hadley) [inline R/dbi-table.R:39]: I think this will fix #737

**Comment by simonpcouch** [comment]
Closes #737.

File: R/dbi-table.R (hunk 2 above)
Comment (detule) [reply R/dbi-table.R]: Thanks, done.

File: R/dbi-table.R (hunk 1 above)
Comment (detule) [reply R/dbi-table.R]: Thanks, done.

File: NEWS.md [hunk 4]
```diff
@@ -35,9 +37,6 @@
 * SQL Server: correctly enumerate schemas across databases in connections pane
   (@detule, #527).
 
-* SQL Server: now uses column type `"BIGINT"` integer64 objects
```
Comment (hadley) [inline NEWS.md:38]: @detule I think you deleted this accidentally?

Comment (detule) [reply NEWS.md:38]: Ugh - sorry about that! Will fix tonight, unless someone beats me to it.

Comment (hadley) [reply NEWS.md:38]: Let me do it since I'm polishing news for release.

File: R/dbi-table.R [hunk 5]
```diff
@@ -152,21 +117,65 @@ setMethod("dbWriteTable", c("OdbcConnection", "SQL", "data.frame"),
...
+      if (!is.null(fieldDetails) && nrow(fieldDetails) == nparam) {
+        result_describe_parameters(rs@ptr, fieldDetails)
+      }
+
+      values <- sqlData(conn, row.names = row.names, value[, , drop = FALSE])
+      if (is.na(batch_rows)) {
+        batch_rows <- NROW(value)
+        if (batch_rows == 0) {
+          batch_rows <- 1
+        }
```
Comment (krlmlr) [inline R/dbi-table.R:158]: Not sure how this condition can ever be `TRUE`, with the check in line 137 in this PR.

File: R/dbi-table.R [hunk 6]
```diff
@@ -152,21 +117,65 @@ setMethod("dbWriteTable", c("OdbcConnection", "SQL", "data.frame"),
...
+    fieldDetails <- tryCatch({
+      details <- odbcConnectionColumns(conn, name, exact = TRUE)
+      details$param_index <- match(details$name, colnames(value))
+      details[!is.na(details$param_index) & !is.na(details$data_type), ]
+    },
+    error = function(e) {
+      return(NULL)
+    })
+
+    if (nrow(value) > 0) {
```
Comment (krlmlr) [inline R/dbi-table.R:137]: Perhaps worth rephrasing as

```r
out <- invisible(NA_real_)

if (nrow(value) == 0) {
  return(out)
}
```

to avoid the indent?

**Review by krlmlr — COMMENTED** [review]
Thanks, great! Do we still need the custom `dbWriteTable()` implementation?

**Pull request merged**